src/lib/io/sol-uart-impl-linux.c

```diff
@@ -32,49 +33,133 @@
 #include "sol-uart.h"
 #include "sol-util-internal.h"
 #include "sol-util-file.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
+#define RX_BUFFER_SIZE (4096)
+
 #define FD_ERROR_FLAGS (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)
 
 struct sol_uart {
+    struct sol_fd *fd_handler;
+    const void *user_data;
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, struct sol_buffer *buf);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    struct sol_timeout *read_timeout;
+    struct sol_ptr_vector pending_blobs;
+    uint8_t rx[RX_BUFFER_SIZE];
+    size_t rx_used;
+    size_t written; //How many bytes We've written so far for the first element of pending_blobs
     int fd;
-    struct {
-        void *rx_fd_handler;
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void *tx_fd_handler;
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length, tx_index;
-    } async;
 };
 
+static void
+dispatch_rx(struct sol_uart *uart)
+{
+    ssize_t r;
+    struct sol_buffer buf = SOL_BUFFER_INIT_CONST(uart->rx, uart->rx_used);
+
+    r = uart->rx_cb((void *)uart->user_data, uart, &buf);
+    SOL_INT_CHECK(r, < 0);
+    SOL_INT_CHECK((size_t)r, > uart->rx_used);
+
+    memmove(uart->rx, uart->rx + uart->rx_used, r);
+    uart->rx_used -= r;
```

Comment:
and if you change that, make the buffer->data malloc(). Since it's 4096 (better to use `sysconf(_SC_PAGESIZE)`), it's likely to be allocated as a page, then if that memory is inline in a structure, it won't fit in a single page... which is worse.